sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/implementation/http/rest/RestProxyBase.java

```diff
@@ -94,51 +92,49 @@ protected abstract Object invoke(Object proxy, Method method, RequestOptions opt
     public abstract void updateRequest(RequestDataConfiguration requestDataConfiguration,
                                        ObjectSerializer objectSerializer) throws IOException;
 
-    @SuppressWarnings({"unchecked", "rawtypes"})
-    public Response createResponse(HttpResponseDecoder.HttpDecodedResponse response, Type entityType,
-                                   Object bodyAsObject) {
-        final Class<? extends Response<?>> cls = (Class<? extends Response<?>>) TypeUtil.getRawClass(entityType);
-        final HttpResponse httpResponse = response.getSourceResponse();
-        final HttpRequest request = httpResponse.getRequest();
-        final int statusCode = httpResponse.getStatusCode();
-        final Headers headers = httpResponse.getHeaders();
-
-        // Inspection of the response type needs to be performed to determine which course of action should be taken to
-        // instantiate the Response<?> from the HttpResponse.
-        //
-        // If the type is either the Response or PagedResponse interface from azure-core a new instance of either
-        // ResponseBase or PagedResponseBase can be returned.
-        if (cls.equals(Response.class)) {
-            // For Response return a new instance of SimpleResponse cast to the class.
-            return cls.cast(new SimpleResponse<>(request, statusCode, headers, bodyAsObject));
-        }
+    @SuppressWarnings({"unchecked"})
+    public Response<?> createResponse(HttpResponseDecoder.HttpDecodedResponse decodedResponse, Type entityType,
+                                      Object bodyAsObject) {
+        final Class<? extends Response<?>> clazz = (Class<? extends Response<?>>) TypeUtil.getRawClass(entityType);
 
-        // Otherwise, rely on reflection, for now, to get the best constructor to use to create the Response subtype.
-        //
-        // Ideally, in the future the SDKs won't need to dabble in reflection here as the Response subtypes should be
-        // given a way to register their constructor as a callback method that consumes HttpDecodedResponse and the body
-        // as an Object.
-        ReflectiveInvoker constructorReflectiveInvoker = RESPONSE_CONSTRUCTORS_CACHE.get(cls);
+        // Inspection of the response type needs to be performed to determine the course of action: either cast the
+        // decoded response to Response or rely on reflection to create an appropriate Response subtype.
+        if (clazz.equals(Response.class)) {
+            if (decodedResponse.getValue() == null) {
+                decodedResponse.setCachedBody(bodyAsObject);
+            }
 
-        return RESPONSE_CONSTRUCTORS_CACHE.invoke(constructorReflectiveInvoker, response, bodyAsObject);
+            // Return the received decoded response cast to Response.
+            return clazz.cast(decodedResponse);
```

Comment:
Wouldnt this throw ClassCastException as its attempting to cast to something that doesn't extend from Response.class?

Reply:
It won't, `decodedResponse` is a subtype of `Response`.

Reply:
Can we add a test case for this?

Reply:
We're not explicitly casting anymore.